services/src/main/java/org/keycloak/storage/ClientStorageManager.java

```diff
@@ -151,18 +151,32 @@ public ClientModel getClientByClientId(RealmModel realm, String clientId) {
 
     @Override
     public Stream<ClientModel> searchClientsByClientIdStream(RealmModel realm, String clientId, Integer firstResult, Integer maxResults) {
-        return query(p -> p.searchClientsByClientIdStream(realm, clientId, firstResult, maxResults), realm, firstResult, maxResults);
+        return query((p, f, m) -> p.searchClientsByClientIdStream(realm, clientId, f, m), realm, firstResult, maxResults);
     }
 
     @Override
     public Stream<ClientModel> searchClientsByAttributes(RealmModel realm, Map<String, String> attributes, Integer firstResult, Integer maxResults) {
-        return query(p -> p.searchClientsByAttributes(realm, attributes, firstResult, maxResults), realm, firstResult, maxResults);
+        return query((p, f, m) -> p.searchClientsByAttributes(realm, attributes, f, m), realm, firstResult, maxResults);
+    }
+
+    @FunctionalInterface
+    interface PaginatedQuery {
+        Stream<ClientModel> query(ClientLookupProvider provider, Integer firstResult, Integer maxResults);
     }
 
-    protected Stream<ClientModel> query(Function<ClientLookupProvider, Stream<ClientModel>> performQuery, RealmModel realm, Integer firstResult, Integer maxResults) {
+    protected Stream<ClientModel> query(PaginatedQuery paginatedQuery, RealmModel realm, Integer firstResult, Integer maxResults) {
         if (maxResults != null && maxResults == 0) return Stream.empty();
 
-        Stream<ClientLookupProvider> providersStream = Stream.concat(Stream.of(session.clientLocalStorage()), getEnabledStorageProviders(session, realm, ClientLookupProvider.class));
+        Set<ClientLookupProvider> extStorageProviders = getEnabledStorageProviders(session, realm, ClientLookupProvider.class).collect(Collectors.toSet());
+        boolean hasExtStorageProviders = !extStorageProviders.isEmpty();
+
+        // when there are external providers involved, we can't do pagination at the lower data layer as we don't know
+        // how many results there will be; i.e. we need to fetch all the clients that match the criteria specified by the query
+        // and perform the pagination later at this layer
+        Integer firstResultLocal = hasExtStorageProviders ? null : firstResult;
```

Comment:
This seems unnecessary. Do you know what was the problem that was causing the failures? Because we are using very similar approach in UserStorageManager, and I believe it works correctly.

Reply:
The problem was that we did the pagination twice – at the data layer and then here. It works for users because we don't do the pagination at the data layer.

Reply:
You can avoid pagination on data layer also here. You just need to put null to first and max parameters.

Reply:
Yes, I know but please see https://github.com/keycloak/keycloak/pull/7870#discussion_r614747022.